File: python/daily_temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: list[int]) -> list[int]:
        """time complexity: O(n)"""
        result: list[int] = [0] * len(temperatures)
        stack: list[tuple[int, int]] = []
        for i, t in enumerate(temperatures):
            while len(stack) > 0 and t > stack[-1][0]:
                _, idx = stack.pop()
                result[idx] = i - idx
            stack.append((t, i))
        return result

    def dailyTemperatures2(self, temperatures: list[int]) -> list[int]:
        """time complexity: O(n)"""
        result: list[int] = [0] * len(temperatures)
        stack: list[int] = []
        for i in range(len(temperatures)):
            while len(stack) > 0 and temperatures[i] > temperatures[stack[-1]]:
                idx: int = stack.pop()
                result[idx] = i - idx
            stack.append(i)
        return result


def test_dailyTemperatures_case_1():
    # arrange
    temperatures: list[int] = [73, 74, 75, 71, 69, 72, 76, 73]
    expected: list[int] = [1, 1, 4, 2, 1, 1, 0, 0]

    # act
    solution = Solution()
    actual = solution.dailyTemperatures(temperatures)

    # assert
    assert actual == expected


def test_dailyTemperatures_case_2():
    # arrange
    temperatures: list[int] = [30, 40, 50, 60]
    expected: list[int] = [1, 1, 1, 0]

    # act
    solution = Solution()
    actual = solution.dailyTemperatures(temperatures)

    # assert
    assert actual == expected


def test_dailyTemperatures_case_3():
    # arrange
    temperatures: list[int] = [30, 60, 90]
    expected: list[int] = [1, 1, 0]

    # act
    solution = Solution()
    actual = solution.dailyTemperatures(temperatures)

    # assert
    assert actual == expected


def test_dailyTemperatures_case_4():
    # arrange
    temperatures: list[int] = [30, 60, 90]
    expected: list[int] = [1, 1, 0]

    # act
    solution = Solution()
    actual = solution.dailyTemperatures2(temperatures)

    # assert
    assert actual == expected

File: rust/src/code/daily_temperatures.rs
struct Solution {}

impl Solution {
    pub fn daily_temperatures(temperatures: Vec<i32>) -> Vec<i32> {
        let mut output: Vec<i32> = vec![0; temperatures.len()];
        let mut stack: Vec<usize> = Vec::new();

        for i in 0..temperatures.len() {
            while let Some(&index) = stack.last() {
                if temperatures[index] < temperatures[i] {
                    stack.pop().unwrap();
                    output[index] = (i - index) as i32;
                } else {
                    break;
                }
            }

            stack.push(i);
        }

        return output;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn daily_temperatures_case_1() {
        // arrange
        let temperatures = vec![73, 74, 75, 71, 69, 72, 76, 73];
        let expected = vec![1, 1, 4, 2, 1, 1, 0, 0];

        // act
        let actual = Solution::daily_temperatures(temperatures);

        assert_eq!(expected, actual);
    }

    #[test]
    fn daily_temperatures_case_2() {
        // arrange
        let temperatures = vec![30, 40, 50, 60];
        let expected = vec![1, 1, 1, 0];

        // act
        let actual = Solution::daily_temperatures(temperatures);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn daily_temperatures_case_3() {
        // arrange
        let temperatures = vec![30, 60, 90];
        let expected = vec![1, 1, 0];

        // act
        let actual = Solution::daily_temperatures(temperatures);

        // assert
        assert_eq!(expected, actual);
    }
}
